README.md

```diff
@@ -1,20 +1,453 @@
+
+Object.defineProperty(pokedex, "pikachu", {
+	configurable: false,
+	writable: true,
+	value: {
+		name: "Pikachu"
+	}
+})
+
+delete pokedex.pikachu // returns false because this property can't be deleted
+pokedex.pikachu = null // since pikachu is no longer being referenced, it has a better chance of being freed from memory
+```
+
+## Enumerating Object Properties
+
+Unlike an array, its not easy to figure out what elements an Object contains. In order to get a value from an Object, we need to know exactly what key to use. Objects don't have a length property and we can't use a normal index based for-loop since the keys are not sequential numbers.
+
+Given the following Object, how could we implement a function that tells us which coupon codes we have left?
+
+```js
+var coupons = {
+	"HOLIDAY": 10,
+	"QHTYB": 50,
+	"MOVIE20": 20
+};
+
+function getCouponCodes(coupons) {
+	// ...
+}
+```
+
+JavaScript provides us with a few ways to do this:
+
+#### for-in loop
+
+A for-in loop allows you to iterate through an Object's keys. As part of the for-in loop, we declare a variable and provide an Object to iterate through. In the example below we're going to iterate through our coupons, storing each key in a new variable we create as part of the loop called `code`.
+
+```js
+for (var code in coupons) {
+	// ... code becomes "HOLIDAY", "QHTYB", or "MOVIE20" after each loop
+}
+```
+
+Inside each loop, a key is chosen from the Object and is assigned to the `code` variable that we specified. We can use this to impement the `getCouponCodes` function:
+
+```js
+function getCouponCodes(coupons) {
+	var codes = [];
+
+	for (var code in coupons) {
+		codes.push(code);
+	}
+	return codes;
+}
+
+getCouponCodes(coupons) // returns ["HOLIDAY", "QHTYB", "MOVIE20"]
+```
+
+Since Objects are an unordered collection of properties, we can't guarantee that the `for-in` loop will always choose keys in the same order they were added to the Object. This function could return the coupon codes in any order.
+
+Now that we have access to the Object's keys, we can use them to get the value for that key. Below is a modified version of `getCouponCodes` that only returns coupon codes who's discount is equal or greater than a certain value.
+
+```js
+function getCouponCodes(coupons, minDiscount) {
+	var codes = [];
+
+	for (var code in coupons) {
+		if (coupons[code] >= minDiscount) {
+			codes.push(code);
+		}
+	}
+	return codes;
+}
+
+getCouponCodes(coupons, 20) // returns ["MOVIE20", "QHTYB"]
+```
+
+We use the key in order to get the value on the Object, which in this case represents the discount for that code.
+
+If there are properties on this Object's prototype chain, they will also be returned in the loop. In most cases this is unwanted, so JavaScript provides a method called `Object.hasOwnProperty` which allows you to verify that the property is a direct property of the Object.
+
+```js
+for (var code in coupons) {
+	if (!coupons.hasOwnProperty(code)) {
+	 	break;
```

Comment:
We want `continue` here so that it behaves as described below, right? ("Adding this check will skip any property that is not a direct property of the Object.") As written, this will skip every property after the first property that isn't an own property of Object.